### PR TITLE
Bug 1846341 - AC automated testing for onProductUrl, requestAnalysis, …

### PR DESCRIPTION
**Pull request by calumozilla**
…requestRecommendations

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1846341

### CODE REVIEW

File: android-components/components/browser/engine-gecko/src/test/java/mozilla/components/browser/engine/gecko/GeckoEngineSessionTest.kt [hunk 1]
```diff
@@ -2544,6 +2544,137 @@ class GeckoEngineSessionTest {
...
+        val ruleResult = GeckoResult<GeckoSession.ReviewAnalysis>()
+        whenever(geckoSession.requestAnalysis("mozilla.com")).thenReturn(ruleResult)
+
+        engineSession.requestProductAnalysis(
+            "mozilla.com",
+            onResult = { onResultCalled = true },
+            onException = { onExceptionCalled = true },
+        )
+
+        // TODO (bug 1855543) need to instantiate ReviewAnalysis
```
Comment (calumozilla) [inline android-components/components/browser/engine-gecko/src/test/java/mozilla/components/browser/engine/gecko/GeckoEngineSessionTest.kt:2581]: Will update this before merging once [bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1855543) is complete

File: android-components/components/browser/engine-gecko/src/test/java/mozilla/components/browser/engine/gecko/GeckoEngineSessionTest.kt [hunk 2]
```diff
@@ -2544,6 +2544,171 @@ class GeckoEngineSessionTest {
...
+                override fun onProductUrlChange(isProductUrl: Boolean) {
+                    productUrlStatus = isProductUrl
+                }
+            },
+        )
+
+        delegate.onProductUrl(geckoSession)
+
+        assertNotNull(productUrlStatus)
+        assertEquals(true, productUrlStatus)
```
Comment (MozillaNoah) [inline android-components/components/browser/engine-gecko/src/test/java/mozilla/components/browser/engine/gecko/GeckoEngineSessionTest.kt:2563]: alternatively:
```suggestion
        assertTrue(productUrlStatus)
```

File: android-components/components/browser/engine-gecko/src/test/java/mozilla/components/browser/engine/gecko/GeckoEngineSessionTest.kt [hunk 3]
```diff
@@ -2544,6 +2544,171 @@ class GeckoEngineSessionTest {
         assertFalse(onExceptionCalled)
     }
 
+    @Test
+    fun `onProductUrlChange should call onSuccess callback for a valid GV response`() {
```
Comment (MozillaNoah) [inline android-components/components/browser/engine-gecko/src/test/java/mozilla/components/browser/engine/gecko/GeckoEngineSessionTest.kt:2548]: nit: let's format the test titles in the GIVEN, WHEN, THEN style

**Review by MozillaNoah — APPROVED** [review]

File: android-components/components/browser/engine-gecko/src/test/java/mozilla/components/browser/engine/gecko/GeckoEngineSessionTest.kt (hunk 3 above)
Comment (calumozilla) [reply android-components/components/browser/engine-gecko/src/test/java/mozilla/components/browser/engine/gecko/GeckoEngineSessionTest.kt]: oh right, re-formatted the test names

**Pull request merged**